keyboards/unicomp/mini_m/minirazz/matrix.c

```diff
@@ -0,0 +1,159 @@
+/* Copyright 2023 Purdea Andrei
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 2 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "quantum.h"
+#include <hardware/spi.h>
+#include <hardware/gpio.h>
+#include <hardware/structs/pads_qspi.h>
+
+matrix_row_t previous_matrix[MATRIX_ROWS];
+
+static inline void setPinInputHigh_atomic(pin_t pin) {
+    ATOMIC_BLOCK_FORCEON {
+        setPinInputHigh(pin);
+    }
+}
+
+static inline void setPinOutput_writeLow(pin_t pin) {
+    ATOMIC_BLOCK_FORCEON {
+        setPinOutput(pin);
+        writePinLow(pin);
+    }
+}
+
+void matrix_init_custom(void) {
+    for (int i=0; i<=7; i++)
+    {
+        setPinInputHigh(i);
+    }
+    for (int i=8; i<=19; i++)
+    {
+        writePinLow(i);
+        setPinInputHigh(i);
+    }
+    spi_init(spi1, 5000000);
+    spi_cpol_t cccpol = 1;
+    spi_cpha_t cccpha = 0;
+    spi_set_format(spi1, 8, cccpol, cccpha, SPI_MSB_FIRST);
+    gpio_set_function(26, GPIO_FUNC_SPI); // SCK
+    gpio_set_function(24, GPIO_FUNC_SPI); // RX (MISO)
```

Comment:
You are correct about the reasons I didn't use the QMK SPI api.
Is this a request for me to look into a core change? I guess it would be possible to change core to put a bunch of `#ifdef` s around any code touching any `SPI_*_PIN` constants. A risk with that would be that someone later could end up defining some defaults for these. Or perhaps we could define a magic value, like -1 meaning that no such pin is assigned, and do `#if SPI_MOSI_PIN != -1`.

Reply:
Unfortunately, things can't be that simple — `SPI_MOSI_PIN` and friends actually have default values at several levels, and although there is the `NO_PIN` define, its value is unsuitable for preprocessor checks (it has a cast to `pin_t` inside, which the preprocessor won't understand; the actual `ioline_t` values for most ChibiOS ports are not suitable for the preprocessor either).  So this might need to be a runtime check for `NO_PIN` instead (although the compiler should be able to optimize out the unused code anyway).

The “no CS” issue might be more involved:
- Currently the `spi_master` code uses the `currentSlavePin != NO_PIN` check to detect whether `spi_start()` has been invoked; this state would need to be decoupled from the pin value.
- The underlying ChibiOS code for `SPI_SELECT_MODE_PAD` does not support `NO_PIN` — it always requires a real (port, pad) pair.  Probably the best way to support a configuration without a real CS pin would be to use `SPI_SELECT_MODE_NONE` in ChibiOS, and then perform any CS pin manipulation on the QMK side; but this requires setting `SPI_SELECT_MODE` in `mcuconf.h`.

Reply:
@sigprof 
I've added these two PRs, to achieve this in core, please review:
https://github.com/qmk/qmk_firmware/pull/20315
https://github.com/qmk/qmk_firmware/pull/20314

I suppose these should be retargetted to develop, right?

I have tested this PR + 20315 + 20314 + the following commit, and it works:
https://github.com/purdeaandrei/qmk_firmware/commit/242d241894b3b82cd7c6b525f11b899b006755bb

I suppose the only annoyance I have is I see in qmk the divisor must be a power of two, but I can live with that for now.

Reply:
This thread should be considered resolved, since I changed this PR to do bitbanging instead, so that it can be merged without waiting for a dependency.